Add S2A AES-GCM implementation and Wycheproof Tests

Ryanfsdf opened the pull request:
Added the initial implementation of S2A AES-GCM as well as the Wycheproof tests. Much of the implementation and some of the tests mirror the AES-GCM implementation in ALTS.

Ryanfsdf commented:
Test Results:
```
=== RUN   TestWycheProofTestVectors
=== RUN   TestWycheProofTestVectors/1/
=== RUN   TestWycheProofTestVectors/2/
=== RUN   TestWycheProofTestVectors/3/
=== RUN   TestWycheProofTestVectors/4/
=== RUN   TestWycheProofTestVectors/5/
=== RUN   TestWycheProofTestVectors/6/
=== RUN   TestWycheProofTestVectors/7/
=== RUN   TestWycheProofTestVectors/8/
=== RUN   TestWycheProofTestVectors/9/
=== RUN   TestWycheProofTestVectors/10/
=== RUN   TestWycheProofTestVectors/11/
=== RUN   TestWycheProofTestVectors/12/
=== RUN   TestWycheProofTestVectors/13/
=== RUN   TestWycheProofTestVectors/14/
=== RUN   TestWycheProofTestVectors/15/
=== RUN   TestWycheProofTestVectors/16/
=== RUN   TestWycheProofTestVectors/17/
=== RUN   TestWycheProofTestVectors/18/
=== RUN   TestWycheProofTestVectors/19/
=== RUN   TestWycheProofTestVectors/20/
=== RUN   TestWycheProofTestVectors/21/special_case
=== RUN   TestWycheProofTestVectors/22/special_case
=== RUN   TestWycheProofTestVectors/23/Flipped_bit_0_in_tag
=== RUN   TestWycheProofTestVectors/24/Flipped_bit_1_in_tag
=== RUN   TestWycheProofTestVectors/25/Flipped_bit_7_in_tag
=== RUN   TestWycheProofTestVectors/26/Flipped_bit_8_in_tag
=== RUN   TestWycheProofTestVectors/27/Flipped_bit_31_in_tag
=== RUN   TestWycheProofTestVectors/28/Flipped_bit_32_in_tag
=== RUN   TestWycheProofTestVectors/29/Flipped_bit_33_in_tag
=== RUN   TestWycheProofTestVectors/30/Flipped_bit_63_in_tag
=== RUN   TestWycheProofTestVectors/31/Flipped_bit_64_in_tag
=== RUN   TestWycheProofTestVectors/32/Flipped_bit_71_in_tag
=== RUN   TestWycheProofTestVectors/33/Flipped_bit_77_in_tag
=== RUN   TestWycheProofTestVectors/34/Flipped_bit_80_in_tag
=== RUN   TestWycheProofTestVectors/35/Flipped_bit_96_in_tag
=== RUN   TestWycheProofTestVectors/36/Flipped_bit_97_in_tag
=== RUN   TestWycheProofTestVectors/37/Flipped_bit_103_in_tag
=== RUN   TestWycheProofTestVectors/38/Flipped_bit_120_in_tag
=== RUN   TestWycheProofTestVectors/39/Flipped_bit_121_in_tag
=== RUN   TestWycheProofTestVectors/40/Flipped_bit_126_in_tag
=== RUN   TestWycheProofTestVectors/41/Flipped_bit_127_in_tag
=== RUN   TestWycheProofTestVectors/42/Flipped_bits_0_and_64_in_tag
=== RUN   TestWycheProofTestVectors/43/Flipped_bits_31_and_63_in_tag
=== RUN   TestWycheProofTestVectors/44/Flipped_bits_63_and_127_in_tag
=== RUN   TestWycheProofTestVectors/45/all_bits_of_tag_flipped
=== RUN   TestWycheProofTestVectors/46/Tag_changed_to_all_zero
=== RUN   TestWycheProofTestVectors/47/tag_changed_to_all_1
=== RUN   TestWycheProofTestVectors/48/msbs_changed_in_tag
=== RUN   TestWycheProofTestVectors/49/lsbs_changed_in_tag
=== RUN   TestWycheProofTestVectors/73/
=== RUN   TestWycheProofTestVectors/74/
=== RUN   TestWycheProofTestVectors/75/
=== RUN   TestWycheProofTestVectors/76/
=== RUN   TestWycheProofTestVectors/77/
=== RUN   TestWycheProofTestVectors/78/
=== RUN   TestWycheProofTestVectors/79/
=== RUN   TestWycheProofTestVectors/80/
=== RUN   TestWycheProofTestVectors/81/
=== RUN   TestWycheProofTestVectors/82/
=== RUN   TestWycheProofTestVectors/83/
=== RUN   TestWycheProofTestVectors/84/
=== RUN   TestWycheProofTestVectors/85/
=== RUN   TestWycheProofTestVectors/86/
=== RUN   TestWycheProofTestVectors/87/
=== RUN   TestWycheProofTestVectors/88/
=== RUN   TestWycheProofTestVectors/89/
=== RUN   TestWycheProofTestVectors/90/
=== RUN   TestWycheProofTestVectors/91/
=== RUN   TestWycheProofTestVectors/92/special_case
=== RUN   TestWycheProofTestVectors/93/special_case
=== RUN   TestWycheProofTestVectors/94/Flipped_bit_0_in_tag
=== RUN   TestWycheProofTestVectors/95/Flipped_bit_1_in_tag
=== RUN   TestWycheProofTestVectors/96/Flipped_bit_7_in_tag
=== RUN   TestWycheProofTestVectors/97/Flipped_bit_8_in_tag
=== RUN   TestWycheProofTestVectors/98/Flipped_bit_31_in_tag
=== RUN   TestWycheProofTestVectors/99/Flipped_bit_32_in_tag
=== RUN   TestWycheProofTestVectors/100/Flipped_bit_33_in_tag
=== RUN   TestWycheProofTestVectors/101/Flipped_bit_63_in_tag
=== RUN   TestWycheProofTestVectors/102/Flipped_bit_64_in_tag
=== RUN   TestWycheProofTestVectors/103/Flipped_bit_71_in_tag
=== RUN   TestWycheProofTestVectors/104/Flipped_bit_77_in_tag
=== RUN   TestWycheProofTestVectors/105/Flipped_bit_80_in_tag
=== RUN   TestWycheProofTestVectors/106/Flipped_bit_96_in_tag
=== RUN   TestWycheProofTestVectors/107/Flipped_bit_97_in_tag
=== RUN   TestWycheProofTestVectors/108/Flipped_bit_103_in_tag
=== RUN   TestWycheProofTestVectors/109/Flipped_bit_120_in_tag
=== RUN   TestWycheProofTestVectors/110/Flipped_bit_121_in_tag
=== RUN   TestWycheProofTestVectors/111/Flipped_bit_126_in_tag
=== RUN   TestWycheProofTestVectors/112/Flipped_bit_127_in_tag
=== RUN   TestWycheProofTestVectors/113/Flipped_bits_0_and_64_in_tag
=== RUN   TestWycheProofTestVectors/114/Flipped_bits_31_and_63_in_tag
=== RUN   TestWycheProofTestVectors/115/Flipped_bits_63_and_127_in_tag
=== RUN   TestWycheProofTestVectors/116/all_bits_of_tag_flipped
=== RUN   TestWycheProofTestVectors/117/Tag_changed_to_all_zero
=== RUN   TestWycheProofTestVectors/118/tag_changed_to_all_1
=== RUN   TestWycheProofTestVectors/119/msbs_changed_in_tag
=== RUN   TestWycheProofTestVectors/120/lsbs_changed_in_tag
--- PASS: TestWycheProofTestVectors (0.01s)
    --- PASS: TestWycheProofTestVectors/1/ (0.00s)
    --- PASS: TestWycheProofTestVectors/2/ (0.00s)
    --- PASS: TestWycheProofTestVectors/3/ (0.00s)
    --- PASS: TestWycheProofTestVectors/4/ (0.00s)
    --- PASS: TestWycheProofTestVectors/5/ (0.00s)
    --- PASS: TestWycheProofTestVectors/6/ (0.00s)
    --- PASS: TestWycheProofTestVectors/7/ (0.00s)
    --- PASS: TestWycheProofTestVectors/8/ (0.00s)
    --- PASS: TestWycheProofTestVectors/9/ (0.00s)
    --- PASS: TestWycheProofTestVectors/10/ (0.00s)
    --- PASS: TestWycheProofTestVectors/11/ (0.00s)
    --- PASS: TestWycheProofTestVectors/12/ (0.00s)
    --- PASS: TestWycheProofTestVectors/13/ (0.00s)
    --- PASS: TestWycheProofTestVectors/14/ (0.00s)
    --- PASS: TestWycheProofTestVectors/15/ (0.00s)
    --- PASS: TestWycheProofTestVectors/16/ (0.00s)
    --- PASS: TestWycheProofTestVectors/17/ (0.00s)
    --- PASS: TestWycheProofTestVectors/18/ (0.00s)
    --- PASS: TestWycheProofTestVectors/19/ (0.00s)
    --- PASS: TestWycheProofTestVectors/20/ (0.00s)
    --- PASS: TestWycheProofTestVectors/21/special_case (0.00s)
    --- PASS: TestWycheProofTestVectors/22/special_case (0.00s)
    --- PASS: TestWycheProofTestVectors/23/Flipped_bit_0_in_tag (0.00s)
    --- PASS: TestWycheProofTestVectors/24/Flipped_bit_1_in_tag (0.00s)
    --- PASS: TestWycheProofTestVectors/25/Flipped_bit_7_in_tag (0.00s)
    --- PASS: TestWycheProofTestVectors/26/Flipped_bit_8_in_tag (0.00s)
    --- PASS: TestWycheProofTestVectors/27/Flipped_bit_31_in_tag (0.00s)
    --- PASS: TestWycheProofTestVectors/28/Flipped_bit_32_in_tag (0.00s)
    --- PASS: TestWycheProofTestVectors/29/Flipped_bit_33_in_tag (0.00s)
    --- PASS: TestWycheProofTestVectors/30/Flipped_bit_63_in_tag (0.00s)
    --- PASS: TestWycheProofTestVectors/31/Flipped_bit_64_in_tag (0.00s)
    --- PASS: TestWycheProofTestVectors/32/Flipped_bit_71_in_tag (0.00s)
    --- PASS: TestWycheProofTestVectors/33/Flipped_bit_77_in_tag (0.00s)
    --- PASS: TestWycheProofTestVectors/34/Flipped_bit_80_in_tag (0.00s)
    --- PASS: TestWycheProofTestVectors/35/Flipped_bit_96_in_tag (0.00s)
    --- PASS: TestWycheProofTestVectors/36/Flipped_bit_97_in_tag (0.00s)
    --- PASS: TestWycheProofTestVectors/37/Flipped_bit_103_in_tag (0.00s)
    --- PASS: TestWycheProofTestVectors/38/Flipped_bit_120_in_tag (0.00s)
    --- PASS: TestWycheProofTestVectors/39/Flipped_bit_121_in_tag (0.00s)
    --- PASS: TestWycheProofTestVectors/40/Flipped_bit_126_in_tag (0.00s)
    --- PASS: TestWycheProofTestVectors/41/Flipped_bit_127_in_tag (0.00s)
    --- PASS: TestWycheProofTestVectors/42/Flipped_bits_0_and_64_in_tag (0.00s)
    --- PASS: TestWycheProofTestVectors/43/Flipped_bits_31_and_63_in_tag (0.00s)
    --- PASS: TestWycheProofTestVectors/44/Flipped_bits_63_and_127_in_tag (0.00s)
    --- PASS: TestWycheProofTestVectors/45/all_bits_of_tag_flipped (0.00s)
    --- PASS: TestWycheProofTestVectors/46/Tag_changed_to_all_zero (0.00s)
    --- PASS: TestWycheProofTestVectors/47/tag_changed_to_all_1 (0.00s)
    --- PASS: TestWycheProofTestVectors/48/msbs_changed_in_tag (0.00s)
    --- PASS: TestWycheProofTestVectors/49/lsbs_changed_in_tag (0.00s)
    --- PASS: TestWycheProofTestVectors/73/ (0.00s)
    --- PASS: TestWycheProofTestVectors/74/ (0.00s)
    --- PASS: TestWycheProofTestVectors/75/ (0.00s)
    --- PASS: TestWycheProofTestVectors/76/ (0.00s)
    --- PASS: TestWycheProofTestVectors/77/ (0.00s)
    --- PASS: TestWycheProofTestVectors/78/ (0.00s)
    --- PASS: TestWycheProofTestVectors/79/ (0.00s)
    --- PASS: TestWycheProofTestVectors/80/ (0.00s)
    --- PASS: TestWycheProofTestVectors/81/ (0.00s)
    --- PASS: TestWycheProofTestVectors/82/ (0.00s)
    --- PASS: TestWycheProofTestVectors/83/ (0.00s)
    --- PASS: TestWycheProofTestVectors/84/ (0.00s)
    --- PASS: TestWycheProofTestVectors/85/ (0.00s)
    --- PASS: TestWycheProofTestVectors/86/ (0.00s)
    --- PASS: TestWycheProofTestVectors/87/ (0.00s)
    --- PASS: TestWycheProofTestVectors/88/ (0.00s)
    --- PASS: TestWycheProofTestVectors/89/ (0.00s)
    --- PASS: TestWycheProofTestVectors/90/ (0.00s)
    --- PASS: TestWycheProofTestVectors/91/ (0.00s)
    --- PASS: TestWycheProofTestVectors/92/special_case (0.00s)
    --- PASS: TestWycheProofTestVectors/93/special_case (0.00s)
    --- PASS: TestWycheProofTestVectors/94/Flipped_bit_0_in_tag (0.00s)
    --- PASS: TestWycheProofTestVectors/95/Flipped_bit_1_in_tag (0.00s)
    --- PASS: TestWycheProofTestVectors/96/Flipped_bit_7_in_tag (0.00s)
    --- PASS: TestWycheProofTestVectors/97/Flipped_bit_8_in_tag (0.00s)
    --- PASS: TestWycheProofTestVectors/98/Flipped_bit_31_in_tag (0.00s)
    --- PASS: TestWycheProofTestVectors/99/Flipped_bit_32_in_tag (0.00s)
    --- PASS: TestWycheProofTestVectors/100/Flipped_bit_33_in_tag (0.00s)
    --- PASS: TestWycheProofTestVectors/101/Flipped_bit_63_in_tag (0.00s)
    --- PASS: TestWycheProofTestVectors/102/Flipped_bit_64_in_tag (0.00s)
    --- PASS: TestWycheProofTestVectors/103/Flipped_bit_71_in_tag (0.00s)
    --- PASS: TestWycheProofTestVectors/104/Flipped_bit_77_in_tag (0.00s)
    --- PASS: TestWycheProofTestVectors/105/Flipped_bit_80_in_tag (0.00s)
    --- PASS: TestWycheProofTestVectors/106/Flipped_bit_96_in_tag (0.00s)
    --- PASS: TestWycheProofTestVectors/107/Flipped_bit_97_in_tag (0.00s)
    --- PASS: TestWycheProofTestVectors/108/Flipped_bit_103_in_tag (0.00s)
    --- PASS: TestWycheProofTestVectors/109/Flipped_bit_120_in_tag (0.00s)
    --- PASS: TestWycheProofTestVectors/110/Flipped_bit_121_in_tag (0.00s)
    --- PASS: TestWycheProofTestVectors/111/Flipped_bit_126_in_tag (0.00s)
    --- PASS: TestWycheProofTestVectors/112/Flipped_bit_127_in_tag (0.00s)
    --- PASS: TestWycheProofTestVectors/113/Flipped_bits_0_and_64_in_tag (0.00s)
    --- PASS: TestWycheProofTestVectors/114/Flipped_bits_31_and_63_in_tag (0.00s)
    --- PASS: TestWycheProofTestVectors/115/Flipped_bits_63_and_127_in_tag (0.00s)
    --- PASS: TestWycheProofTestVectors/116/all_bits_of_tag_flipped (0.00s)
    --- PASS: TestWycheProofTestVectors/117/Tag_changed_to_all_zero (0.00s)
    --- PASS: TestWycheProofTestVectors/118/tag_changed_to_all_1 (0.00s)
    --- PASS: TestWycheProofTestVectors/119/msbs_changed_in_tag (0.00s)
    --- PASS: TestWycheProofTestVectors/120/lsbs_changed_in_tag (0.00s)
=== RUN   TestAESGCMEncrypt
--- PASS: TestAESGCMEncrypt (0.00s)
=== RUN   TestAESGCMEncryptRoundtrip
--- PASS: TestAESGCMEncryptRoundtrip (0.00s)
=== RUN   TestAESGCMInvalidKeySize
--- PASS: TestAESGCMInvalidKeySize (0.00s)
PASS
ok      google.golang.org/grpc/security/s2a/internal/crypter    0.032s
```

matthewstevenson88 commented on security/s2a/internal/crypter/aesgcm.go:
```diff
@@ -0,0 +1,97 @@
+
+// Supported key sizes in bytes.
+const (
+	AesGcmKeySize128 = 16
```
I think the first letter of the constants should be lower case, since we don't want these constants to be exported. (Cesar may correct me on Go style.)

matthewstevenson88 replied:
Can you change the name formats to `aes128GcmKeySize` and `aes256GcmKeySize`? This will make it consistent across all of the S2A code.

matthewstevenson88 commented on security/s2a/internal/crypter/aesgcm.go:
```diff
@@ -0,0 +1,97 @@
+}
+
+func (s *aesgcm) UpdateKey(key []byte) error {
+	if len(key) != AesGcmKeySize128 && len(key) != AesGcmKeySize256 {
```
The key update should not be allowed to change the key size, so please make sure that the new key has the same length as the old key.

matthewstevenson88 commented on security/s2a/internal/crypter/aesgcm.go:
```diff
@@ -0,0 +1,97 @@
+	AesGcmKeySize256 = 32
+)
+
+// aesgcm is the struct that holds necessary information for S2A record.
```
Could we change this to e.g. `aesgcm is the struct that holds an AES-GCM cipher for the S2A AEAD crypter.`?

matthewstevenson88 commented on security/s2a/internal/crypter/aesgcm.go:
```diff
@@ -0,0 +1,97 @@
+	aead cipher.AEAD
+}
+
+// NewAESGCM creates an instance that uses aesgcm for S2A record. Note that the
```
Can we change `an instance that uses aesgcm for S2A record` to `an aesgcm implementation of the S2AAeadCrypter interface`?

Ryanfsdf replied:
How does `NewAESGCM creates an instance of an AES-GCM crypter of the S2AAeadCrypter interface.` sound? I'm hesitant about `NewAESGCM creates an aesgcm implementation of the S2AAeadCrypter interface` since we're creating an instance of the class rather than an implementation. WDYT?

matthewstevenson88 replied:
What about if we just left `NewAESGCM creates an AES-GCM crypter instance.`? (Having `of the S2AAeadCrypter interface` at the end doesn't read so well.)

matthewstevenson88 commented on security/s2a/internal/crypter/aesgcm_test.go:
```diff
@@ -0,0 +1,319 @@
+)
+
+// cryptoTestVector is a struct for a AES-GCM test vector
+type cryptoTestVector struct {
```
Sorry I forgot to mention this earlier. We also need the NIST and IEEE test vectors (see [here](https://github.com/grpc/grpc-java/blob/62620ccd00b4bc772ba6622a7c1b192dea45747c/alts/src/test/java/io/grpc/alts/internal/AesGcmHkdfAeadCrypterTest.java#L156)). There are just a few of them, so they should be reasonable to hardcode as test vectors.

matthewstevenson88 commented on security/s2a/internal/crypter/aesgcm_test.go:
```diff
@@ -0,0 +1,319 @@
+	}
+}
+
+// Test encrypt and decrypt using an invalid key size.
```
Can you add a test where we 1. create the crypter, 2. update the key, 3. use the new key to encrypt/decrypt. And also a test where we create the crypter, then update the key but it fails.

matthewstevenson88 commented:
Thanks Ryan! The PR is looking great - I left a few comments.

One general thought: would it be possible to spin the JSON parsing out of the unit test and into a separate `test_util` directory? We can discuss this at today's standup if you like.

Ryanfsdf commented:
Thanks for the review! I addressed all the review comments. 

I refactored out the wycheproof testing utils into the `testutil` directory. `ParseWycheProofTestVectors` takes a path to the wycheproof test vectors and also a function to filter out unsupported test vectors. Davis will need need to implement a filter function for testing his CHACHA-POLY implementation to pass into `ParseWycheProofTestVectors` (it could possibly just be an empty function, I'm not sure how CHACHA-POLY will be implemented). This should be generic enough to work with any AEAD crypter implementation which uses wycheproof test vectors.

PTAL!

matthewstevenson88 commented on security/s2a/internal/crypter/aesgcm_test.go:
```diff
@@ -307,13 +231,252 @@ func TestAESGCMInvalidKeySize(t *testing.T) {
+	// http://www.ieee802.org/1/files/public/docs2011/bn-randall-test-vectors-0511-v1.pdf
+	for _, test := range []CryptoTestVector{
+		{
+			Comment:    "Derived from NIST test vector 1",
```
Aren't these the actual NIST/IEEE test vectors, and not just a test vector derived from them?

Ryanfsdf replied:
Yes I believe so. The comment was from the Java/C++ NIST/IEEE unit tests which I converted to Go. I left the comment as it originally was in the Java version. I'll remove the `Derived from`.

matthewstevenson88 replied:
SGTM. Thanks!

matthewstevenson88 commented on security/s2a/internal/crypter/aesgcm_test.go:
```diff
@@ -307,13 +231,252 @@ func TestAESGCMInvalidKeySize(t *testing.T) {
+	}
+}
+
+// Test update key for AES-GCM using a key with an invalid size.
```
I'm not sure this comment is correct?

matthewstevenson88 commented on security/s2a/internal/crypter/aesgcm_test.go:
```diff
@@ -80,176 +38,142 @@ func getGCMCryptoPair(key []byte, t *testing.T) (S2AAeadCrypter, S2AAeadCrypter)
 	}
 }
 
 // Test encrypt and decrypt using test vectors for aes128gcm.
```
Where are these test vectors coming from?

matthewstevenson88 commented on security/s2a/internal/crypter/aesgcm_test.go:
```diff
@@ -80,176 +38,142 @@ func getGCMCryptoPair(key []byte, t *testing.T) (S2AAeadCrypter, S2AAeadCrypter)
+		testGroup.TagSize != 128
 }
 
 func TestWycheProofTestVectors(t *testing.T) {
```
Should this function start with a lowercase `t` rather than `T`? I don't think we want this function to be used outside of this file, correct?

Ryanfsdf replied:
Go unit tests should start with the keyword `Test`(with a capital T) for the testing framework to recognize it as a unit test.

matthewstevenson88 replied:
Good point. We shall leave it as is then :)

matthewstevenson88 commented on security/s2a/internal/crypter/aesgcm_test.go:
```diff
@@ -20,53 +20,11 @@ package crypter
-	TestGroups []testGroup `json:"testGroups"`
-}
-
 // getGCMCryptoPair outputs a client/server pair on AES-GCM.
```
Can we call these something else than `client` and `server`? Perhaps `protector` and `unprotector`, or `sender` and `receiver`?

Ryanfsdf replied:
I took the naming from the ALTS unit tests. I think the consistency of the naming between this and ALTS would be better kept the same, WYDT?

matthewstevenson88 replied:
The `client` and `server` naming in this context implies that the client does the encrypting and the server does the decrypting, which is not the case - both do both. That's why I was suggesting the different naming. I'm not sure why ALTS chose that naming scheme, but I don't think we should keep it just for the sake of consistency.

Ryanfsdf replied:
Got it, will change it then!

matthewstevenson88 commented on security/s2a/internal/crypter/aesgcm_test.go:
```diff
@@ -290,7 +214,7 @@ func testGCMEncryptRoundtrip(client S2AAeadCrypter, server S2AAeadCrypter, t *te
-	for _, keySize := range []int{AesGcmKeySize128, AesGcmKeySize256} {
+	for _, keySize := range []int{aes128GcmKeySize, aes256GcmKeySize} {
 		key := make([]byte, keySize)
 		client, server := getGCMCryptoPair(key, t)
```
Similarly here: please change `client` and `server`.

matthewstevenson88 commented on security/s2a/internal/crypter/testutil/common.go:
```diff
@@ -0,0 +1,54 @@
+	InvalidResult = "invalid"
+)
+
+// CryptoTestVector is a struct for a test vector
```
Please mention that this is a test vector for an `S2AAeadCrypter`  instance.

matthewstevenson88 commented:
Thanks Ryan! The PR is looking good, and the Wycheproof parsing that you set up will be very useful when @davisgu writes the CHACHA-POLY implementation.

I left a few more comments, and had one general comment on the structure of the test. Could you organize the code  so that all similar functions are together? For example, first are all setup functions, then the tests for error behavior, and then all the roundtrip tests at the end. (That's just one suggested order, feel free to group them however you think is best.)

Ryanfsdf commented on security/s2a/internal/crypter/aesgcm.go:
```diff
@@ -26,17 +26,21 @@ import (
 
 // Supported key sizes in bytes.
```
Should we call it `aes16GcmKeySize` and `aes32GcmKeySize`? I left it as 128/256 since I've never seen it referred to in bytes rather than bits. Using 16/32 might be a bit unconventional, WDYT?

matthewstevenson88 replied:
Good catch! I think these are the only places where we should keep 128/256, because AES-128-GCM and AES-256-GCM are recognized names.

Ryanfsdf commented:
Thanks for the review! I've addressed all the review comments and reorganized the test file. PTAL!

matthewstevenson88 commented on security/s2a/internal/crypter/common.go:
```diff
@@ -0,0 +1,29 @@
+	// GcmTagSize is the tag size - the difference in length between plaintext
+	// and ciphertext. From crypto/cipher/gcm.go in the Go crypto library.
+	GcmTagSize = 16
+	// NonceSize is the size of the nonce in number of bytes.
```
Can you mention that this is the nonce size for AES-128-GCM-SHA256, AES-256-GCM-SHA384, and CHACHA20-POLY1305-SHA256? (It might not be the nonce size for a ciphersuite that we add in the future.)

matthewstevenson88 commented on security/s2a/internal/crypter/aesgcm.go:
```diff
@@ -0,0 +1,105 @@
+	return dst, nil
+}
+
+func (s *aesgcm) Decrypt(dst, ciphertext, nonce, aad []byte) ([]byte, error) {
```
Please check that `len(nonce)` is equal to `NonceSize`.

matthewstevenson88 commented on security/s2a/internal/crypter/aesgcm.go:
```diff
@@ -0,0 +1,105 @@
+// has enough capacity to hold these bytes, the ciphertext and the tag, no
+// allocation and copy operations will be performed. dst and plaintext may
+// fully overlap or not at all.
+func (s *aesgcm) Encrypt(dst, plaintext, nonce, aad []byte) ([]byte, error) {
```
Please check that `len(nonce)` is equal to `NonceSize`.

matthewstevenson88 commented on security/s2a/internal/crypter/common.go:
```diff
@@ -0,0 +1,29 @@
+
+const (
+	// GcmTagSize is the tag size - the difference in length between plaintext
+	// and ciphertext. From crypto/cipher/gcm.go in the Go crypto library.
```
Please remove `the difference in length between plaintext and ciphertext`. (This will be true at the AEAD crypter layer but not at the record protocol layer. Since this is file is not AEAD crypter specific, I want to avoid confusion.)

matthewstevenson88 replied:
Please also add `the tag size in bytes`.

matthewstevenson88 commented on security/s2a/internal/crypter/common.go:
```diff
@@ -0,0 +1,29 @@
+	NonceSize = 12
+)
+
+// ErrAuth is the error produced on authentication failure.
```
Can you make the comment more precise? This is the error produced by the AEAD crypter when something happens.

Ryanfsdf replied:
I just removed this and added the message directly in `Decrypt` since that is the only place where it is used. I had it initially since it was present in ALTS.

matthewstevenson88 commented:
Thanks for the PR Ryan! I left a few minor comments and approved. Adding @cesarghali for Go readability.

matthewstevenson88 commented on security/s2a/internal/crypter/aesgcm_test.go:
```diff
@@ -20,53 +20,11 @@ package crypter
-	TestGroups []testGroup `json:"testGroups"`
-}
-
 // getGCMCryptoPair outputs a client/server pair on AES-GCM.
```
nit: client/server

Ryanfsdf replied:
I think you may be looking at a previous revision.

matthewstevenson88 commented on security/s2a/internal/crypter/aesgcm_test.go:
```diff
@@ -20,53 +20,11 @@ package crypter
-
 // getGCMCryptoPair outputs a client/server pair on AES-GCM.
 func getGCMCryptoPair(key []byte, t *testing.T) (S2AAeadCrypter, S2AAeadCrypter) {
 	client, err := NewAESGCM(key)
```
nit: client

matthewstevenson88 commented:
Also please remove all instances of `client` and `server` from the tests, not just the couple I commented on.

Ryanfsdf commented:
> Also please remove all instances of `client` and `server` from the tests, not just the couple I commented on.

I think you may have been looking a previous revision, I've removed all occurences of `client` and `server` from the test file I believe.

Ryanfsdf commented:
Addressed all the review comments. I also added unit tests for checking invalid nonce sizes, since I added a nonce size check in `Encrypt` and `Decrypt`. PTAL!

cesarghali commented on security/s2a/internal/crypter/testutil/common.go:
```diff
@@ -0,0 +1,55 @@
+
+// TestVector is a struct for a WycheProof test vector.
+type TestVector struct {
+	TcId    int    `json:"tcId"`
```
Please call this `TcID`:
https://github.com/golang/go/wiki/CodeReviewComments#initialisms

Ryanfsdf replied:
Done

Ryanfsdf commented on security/s2a/internal/crypter/testutil/wycheproofutil.go:
```diff
@@ -0,0 +1,54 @@
+
+	var tf TestFile
+	err = dec.Decode(&tf)
+	if err != nil {
```
`ValidResult` and `InvalidResult` will still need to be exported since it is used throughout the `aesgcm_test.go` file.

Ryanfsdf commented on security/s2a/internal/crypter/testutil/wycheproofutil.go:
```diff
@@ -0,0 +1,54 @@
+// ParseWycheProofTestVectors takes a path to a WycheProof test vector, a test
+// group filter, and returns the resulting CryptoTestVector. The test group
+// filter will be used to filter out unsupported test inputs.
+func ParseWycheProofTestVectors(
```
Got it, I was debating on using a dot import here or not and I initially decided to use it since otherwise we'll have to write `testutil` all across the test. I updated it to use the fully qualified name now.

cesarghali replied:
SG, thanks for the fix!

Ryanfsdf commented on security/s2a/internal/crypter/aesgcm.go:
```diff
@@ -0,0 +1,114 @@
+// either 128 bits or 256 bits.
+func NewAESGCM(key []byte) (S2AAeadCrypter, error) {
+	if len(key) != aes128GcmKeySize && len(key) != aes256GcmKeySize {
+		return nil, fmt.Errorf("supplied key must be 128 or 256 bits. given: %d", len(key)*8)
```
Done

Ryanfsdf commented on security/s2a/internal/crypter/aesgcm.go:
```diff
@@ -0,0 +1,114 @@
+	if len(nonce) != NonceSize {
+		return nil, fmt.Errorf("nonce size must be %d bytes. received: %d",
+			NonceSize,
+			len(nonce))
```
Done

Ryanfsdf commented on security/s2a/internal/crypter/aesgcm.go:
```diff
@@ -0,0 +1,114 @@
+			len(nonce))
+	}
+	// If we need to allocate an output buffer, we want to include space for
+	// GCM tag to avoid forcing S2A record to reallocate as well.
```
Yes, this part was taken from ALTS (I just swapped ALTS for S2A). I'll make equivalent changes for ALTS.

Ryanfsdf commented on security/s2a/internal/crypter/aesgcm.go:
```diff
@@ -0,0 +1,114 @@
+	if len(nonce) != NonceSize {
+		return nil, fmt.Errorf("nonce size must be %d bytes. received: %d",
+			NonceSize,
+			len(nonce))
```
Done

Ryanfsdf commented on security/s2a/internal/crypter/common.go:
```diff
@@ -0,0 +1,25 @@
+const (
+	// GcmTagSize is the tag size in bytes. From crypto/cipher/gcm.go in the Go
+	// crypto library.
+	GcmTagSize = 16
```
Done, and made equivalent changes in ALTS.

Ryanfsdf commented on security/s2a/internal/crypter/common.go:
```diff
@@ -0,0 +1,25 @@
+	GcmTagSize = 16
+	// NonceSize is the size of the nonce in number of bytes for
+	// AES-128-GCM-SHA256, AES-256-GCM-SHA384, and CHACHA20-POLY1305-SHA256.
+	NonceSize = 12
```
Done

Ryanfsdf commented on security/s2a/internal/crypter/aesgcm_test.go:
```diff
@@ -0,0 +1,513 @@
+import (
+	"bytes"
+	"fmt"
+	. "google.golang.org/grpc/security/s2a/internal/crypter/testutil"
```
Got it, I was debating on using a dot import here or not and I initially decided to use it since otherwise we'll have to write `testutil` all across the test. I updated it to use the fully qualified name now.

cesarghali replied:
SG, thanks!

Ryanfsdf commented on security/s2a/internal/crypter/testutil/common.go:
```diff
@@ -0,0 +1,55 @@
+// Constants indicating whether the test vector is valid or not.
+const (
+	ValidResult   = "valid"
+	InvalidResult = "invalid"
```
They're used throughout `aesgcm_test.go` so they can't be unexported.

cesarghali replied:
Ah, right. Let's keep them exported then.

Ryanfsdf commented on security/s2a/internal/crypter/aesgcm_test.go:
```diff
@@ -0,0 +1,513 @@
+	return sender, receiver
+}
+
+func isFailure(result string, err error, got, expected []byte) bool {
```
They're used in other places throughout the test so they still need to be exported.

Ryanfsdf commented on security/s2a/internal/crypter/aesgcm_test.go:
```diff
@@ -0,0 +1,513 @@
+	// Ciphertext is: encrypted text + tag.
+	var ciphertext []byte
+	ciphertext = append(ciphertext, test.Ciphertext...)
+	ciphertext = append(ciphertext, test.Tag...)
```
Yes that works. Done

Ryanfsdf commented on security/s2a/internal/crypter/aesgcm_test.go:
```diff
@@ -0,0 +1,513 @@
+		testGroup.TagSize != 128
+}
+
+func testGCMEncryptionDecryption(sender S2AAeadCrypter, receiver S2AAeadCrypter, test *CryptoTestVector, t *testing.T) {
```
No specific reason, I agree that it makes more sense to do encryption before decryption with how the test is named. I updated it to do encryption first and changed the equivalent test in ALTS which this was taken from.

Ryanfsdf commented on security/s2a/internal/crypter/aesgcm_test.go:
```diff
@@ -0,0 +1,513 @@
+	const plaintext = "This is plaintext."
+	var err error
+	buf := []byte(plaintext)
+	buf, err = sender.Encrypt(buf[:0], buf, nonce, nil)
```
I don't think we want to do this as this will be testing a different part of the code than what we want. Here, are reusing the buffer as both the input and the output (hence, why buf represents both ciphertext as well as plaintext throughout it's lifetime). I think this is intended to test the case where the input fully overlaps with the output. WDYT?

cesarghali replied:
This is fine. Thanks for the clarification. Can you please document this in a comment so we don't forget about it in the future?

Ryanfsdf replied:
Done

Ryanfsdf commented on security/s2a/internal/crypter/aesgcm_test.go:
```diff
@@ -0,0 +1,513 @@
+	const plaintext = "This is plaintext."
+	var err error
+	buf := []byte(plaintext)
+	buf, err = sender.Encrypt(buf[:0], buf, nonce, nil)
```
Same comment as above.

Ryanfsdf commented on security/s2a/internal/crypter/aesgcm_test.go:
```diff
@@ -0,0 +1,513 @@
+
+	// Decrypt first message.
+	ciphertext := append([]byte(nil), buf...)
+	buf, err = receiver.Decrypt(buf[:0], buf, nonce, nil)
```
> s/buf/decryptedPlaintext

Done

> or something similar. I'm proposing variable name changes because buf is not a clear name of what it holds. Also, it's meaning is changing in the code, from plaintext to ciphertext to plaintext again.

This was originally taken from ALTS and I think it is intentional. `buf` represents both ciphertext and plaintext at different times throughout its lifetime because we use `buf` as both the input and output buffer. This is for testing the case where the input and output fully overlap.

Ryanfsdf commented on security/s2a/internal/crypter/aesgcm_test.go:
```diff
@@ -0,0 +1,513 @@
+	// Use 17 bytes, which is invalid
+	key := make([]byte, 17)
+	_, err := NewAESGCM(key)
+	if err == nil {
```
Done

Ryanfsdf commented on security/s2a/internal/crypter/aesgcm_test.go:
```diff
@@ -0,0 +1,513 @@
+	key := make([]byte, aes128GcmKeySize)
+	crypter, err := NewAESGCM(key)
+	if err != nil {
+		t.Fatalf("NewAESGCM(key) = %v", err)
```
Done

Ryanfsdf commented on security/s2a/internal/crypter/aesgcm_test.go:
```diff
@@ -0,0 +1,513 @@
+	key := make([]byte, aes128GcmKeySize)
+	crypter, err := NewAESGCM(key)
+	if err != nil {
+		t.Fatalf("NewAESGCM(key) = %v", err)
```
Done

Ryanfsdf commented on security/s2a/internal/crypter/testutil/common.go:
```diff
@@ -0,0 +1,55 @@
+	TcId    int    `json:"tcId"`
+	Comment string `json:"comment"`
+	Key     string `json:"key"`
+	Iv      string `json:"iv"`
```
Done

Ryanfsdf commented on security/s2a/internal/crypter/testutil/common.go:
```diff
@@ -0,0 +1,55 @@
+// CryptoTestVector is a struct representing a test vector for an S2AAeadCrypter
+// instance.
+type CryptoTestVector struct {
+	Id                                          int
```
Done

Ryanfsdf commented on security/s2a/internal/crypter/aesgcm_test.go:
```diff
@@ -0,0 +1,513 @@
+	}
+
+	// Decrypt first message.
+	ciphertext := append([]byte(nil), buf...)
```
Same as below.

Ryanfsdf commented on security/s2a/internal/crypter/testutil/common.go:
```diff
@@ -0,0 +1,55 @@
+	Iv      string `json:"iv"`
+	Aad     string `json:"aad"`
+	Msg     string `json:"msg"`
+	Ct      string `json:"ct"`
```
It's a shortcut for ciphertext, so I'll leave it as is.

Ryanfsdf commented on security/s2a/internal/crypter/testutil/common.go:
```diff
@@ -0,0 +1,55 @@
+
+// TestGroup is a struct for a WycheProof test group.
+type TestGroup struct {
+	IvSize  int          `json:"ivSize"`
```
Done

Ryanfsdf commented on security/s2a/internal/crypter/testutil/common.go:
```diff
@@ -0,0 +1,55 @@
+	}
+	b, err := hex.DecodeString(s)
+	if err != nil {
+		panic(err)
```
I think not using a panic here might make the tests more difficult to read.
Currently, we construct our test cases like this:
```
{
			Key:         testutil.Dehex("5b9604fe14eadba931b0ccf34843dab9"),
			Plaintext:   testutil.Dehex("001d0c231287c1182784554ca3a21908"),
			Ciphertext:  testutil.Dehex("26073cc1d851beff176384dc9896d5ff"),
			Tag:         testutil.Dehex("0a3ea7a5487cb5f7d70fb6c58d038554"),
			Nonce:       testutil.Dehex("028318abc1824029138141a2"),
			Result:      testutil.ValidResult,
			AllocateDst: true,
		},
```
If we return an error on Dehex, we would need to make this much longer by declaring and error checking every use of Dehex. I think panics should be avoided in real production code but I think they may be useful sometimes, particular in cases like this where the function is only used for testing.

The dehex method in ALTS also panics, and is also only used for unit tests.

cesarghali replied:
You're right. It's totally fine to use `panic` here. Since using `panic` is not very common, I wanted to take the opportunity to mention (as FYI) the Golang style guide rule of not using `panic` if possible :)

Ryanfsdf commented on security/s2a/internal/crypter/aesgcm_test.go:
```diff
@@ -0,0 +1,513 @@
+/*
```
Done, I tried to combine all the ones I could combine.

Ryanfsdf commented on security/s2a/internal/crypter/aesgcm_test.go:
```diff
@@ -0,0 +1,513 @@
+
+// Test update key for AES-GCM using a key with different size from the initial
+// key.
+func TestAESGCMMismatchKeySizeUpdate(t *testing.T) {
```
Done

Ryanfsdf commented on security/s2a/internal/crypter/aesgcm_test.go:
```diff
@@ -0,0 +1,513 @@
+	// Decrypt first message.
+	ciphertext := append([]byte(nil), buf...)
+	buf, err = receiver.Decrypt(buf[:0], buf, nonce, nil)
+	if err != nil || string(buf) != plaintext {
```
Done

Ryanfsdf commented on security/s2a/internal/crypter/aesgcm_test.go:
```diff
@@ -0,0 +1,513 @@
+	// Decrypt.
+	got, err := receiver.Decrypt(nil, ciphertext, test.Nonce, test.Aad)
+	if isFailure(test.Result, err, got, test.Plaintext) {
+		t.Errorf("key=%v\ntag=%v\nciphertext=%v\nDecrypt = %v, %v\nwant: %v",
```
Done

Ryanfsdf commented on security/s2a/internal/crypter/aesgcm_test.go:
```diff
@@ -0,0 +1,513 @@
+	}
+	got, err = sender.Encrypt(dst[:0], test.Plaintext, test.Nonce, test.Aad)
+	if isFailure(test.Result, err, got, ciphertext) {
+		t.Errorf("key=%v\nplaintext=%v\nEncrypt = %v, %v\nwant: %v",
```
Done

Ryanfsdf commented on security/s2a/internal/crypter/aesgcm_test.go:
```diff
@@ -0,0 +1,513 @@
+	buf := []byte(plaintext)
+	buf, err = sender.Encrypt(buf[:0], buf, nonce, nil)
+	if err != nil {
+		t.Fatal("Encrypting with sender-side context: unexpected error", err, "\n",
```
Done

Ryanfsdf commented on security/s2a/internal/crypter/aesgcm_test.go:
```diff
@@ -0,0 +1,513 @@
+
+	// Decryption fails: replay attack.
+	if got, err := receiver.Decrypt(nil, buf, nonce, nil); err == nil {
+		t.Error("Decrypting sender-side ciphertext with a sender-side context unexpectedly succeeded; want unexpected counter error:\n",
```
Removed this, as it isn't necessary for this test. I initially took it from ALTS but this test was used to test the counter, which isn't present in this AES-GCM implemetation.

Ryanfsdf commented:
Thanks for the reviews! Addressed all the review comments, PTAL.

cesarghali commented on security/s2a/internal/crypter/aesgcm_test.go:
```diff
@@ -0,0 +1,474 @@
+}
+
+func wycheProofTestVectorFilter(testGroup testutil.TestGroup) bool {
+	return testGroup.IVSize != 96 ||
```
Please add a comment for why we're filtering out these IV, Key and Tag sizes.

Ryanfsdf replied:
Done

cesarghali commented on security/s2a/internal/crypter/aesgcm_test.go:
```diff
@@ -0,0 +1,474 @@
+}
+
+func testGCMEncryptionDecryption(sender S2AAeadCrypter, receiver S2AAeadCrypter, test *testutil.CryptoTestVector, t *testing.T) {
+	// Ciphertext is: encrypted text + tag.
```
Please use `ciphertext`, with a small `c`, to make it clear that it's referring to the new variable we're defining in the next line.

Ryanfsdf replied:
Done

cesarghali commented on security/s2a/internal/crypter/aesgcm_test.go:
```diff
@@ -0,0 +1,474 @@
+}
+
+// Test Encrypt/Decrypt using an invalid nonce size.
+func TestAESGCMEncryptDecryptInvalidNonce(t *testing.T) {
```
The `TestAESGCMEncrypt` below seems to attempt to encrypt and decrypt successfully. Can we merge this test into the one below? Basically add a new test case with an invalid nonce.

Ryanfsdf replied:
Done

cesarghali commented on security/s2a/internal/crypter/aesgcm_test.go:
```diff
@@ -0,0 +1,474 @@
+			key := make([]byte, keySize)
+			sender, receiver := getGCMCryptoPair(key, t)
+			if shouldUpdateKeys {
+				// Update the key with a new one which is different from the original.
```
This might be already the case, but please make sure this comment line is not longer than 80 chars.

Ryanfsdf replied:
Done

cesarghali commented on security/s2a/internal/crypter/aesgcm_test.go:
```diff
@@ -0,0 +1,474 @@
+		for _, keySize := range []int{aes128GcmKeySize, aes256GcmKeySize} {
+			key := make([]byte, keySize)
+			sender, receiver := getGCMCryptoPair(key, t)
+			if shouldUpdateKeys {
```
How about we try to call `testGCMEncryptRoundtrip(sender, receiver, t)` once before this line? Basically, we will encrypt/decrypt, update keys if needed, and then encrypt/decrypt again. WDYT?

Ryanfsdf replied:
That sounds good. Done.

cesarghali commented on security/s2a/internal/crypter/aesgcm_test.go:
```diff
@@ -0,0 +1,474 @@
+}
+
+// Test encrypt and decrypt using test vectors for aes128gcm. Much of these
+// test vectors were taken from the ALTS AES-GCM unit tests.
```
Please remove any reference to ALTS in the code. You can delete the second sentence of this comment.

Ryanfsdf replied:
Done. I added the second sentence due to a previous review comment from Matt, I think it makes sense to remove any reference to ALTS though.

cesarghali commented on security/s2a/internal/crypter/aesgcm_test.go:
```diff
@@ -0,0 +1,474 @@
+		},
+		{
+			Key:         testutil.Dehex("11754cd72aec309bf52f7687212e8957"),
+			Plaintext:   nil,
```
You can delete any field that has the default value. Some default values are: nil, 0, "", false.

Ryanfsdf replied:
Done, though I think some explicit values may help with determining what exact case the test vector is testing.

cesarghali replied:
This is why we added `Comment` :)

cesarghali commented on security/s2a/internal/crypter/aesgcm_test.go:
```diff
@@ -0,0 +1,474 @@
+// Test encrypt and decrypt using test vectors for aes128gcm. Much of these
+// test vectors were taken from the ALTS AES-GCM unit tests.
+func TestAESGCMEncrypt(t *testing.T) {
+	for _, test := range []testutil.CryptoTestVector{
```
Please use the `Comment` field here as well and pass it to `t.Run`. See my comment below about that.

cesarghali commented on security/s2a/internal/crypter/aesgcm_test.go:
```diff
@@ -0,0 +1,474 @@
+		"testdata/aes_gcm_wycheproof.json",
+		wycheProofTestVectorFilter,
+		t,
+	) {
```
Please combine lines 267-271 into a single line.

Ryanfsdf replied:
Done

cesarghali commented on security/s2a/internal/crypter/aesgcm_test.go:
```diff
@@ -0,0 +1,474 @@
+			AllocateDst: true,
+		},
+	} {
+		sender, receiver := getGCMCryptoPair(test.Key, t)
```
Should we use `t.Run` here as well?

Ryanfsdf replied:
Yes, with the addition of `Comment`, I think that makes sense.

cesarghali commented on security/s2a/internal/crypter/aesgcm_test.go:
```diff
@@ -132,114 +135,79 @@ func TestAESGCMKeySizeUpdate(t *testing.T) {
-				}
-			}
-			testGCMEncryptRoundtrip(sender, receiver, t)
+	for _, keySize := range []int{aes128GcmKeySize, aes256GcmKeySize} {
```
Sorry for not clarifying my comment. Let's keep the `for _, shouldUpdateKeys := range []bool{true, false}` loop.

Ryanfsdf replied:
I don't think the `shouldUpdateKeys` loop will be useful here since we don't change anything between each roundtrip call. Testing back-to-back roundtrips in the S2A Half Connection will be useful since the nonce will be updated with each roundtrip, but it wouldn't be useful here since we don't update the nonce.
